<h1>Lab | Revisiting Machine Learning Case Study</h1><br>

In this lab, you will use learningSet.csv file which you already have cloned in today's activities.

<h2>Instructions</h2><br>

Complete the following steps on the categorical columns in the dataset:

* 1. Check for null values in all the columns<br>
<br>
* 2. Exclude the following variables by looking at the definitions. Create a new empty list called drop_list. We will append this list and then drop all the columns in this list later:

        - OSOURCE - symbol definitions not provided, too many categories
        - ZIP CODE - we are including state already<br>
<br>
* 3. Identify columns that over 85% missing values<br>
<br>
* 4. Remove those columns from the dataframe<br>
<br>
* 5. Reduce the number of categories in the column GENDER. The column should only have either "M" for males, "F" for females, and "other" for all the rest

    - Note that there are a few null values in the column. We will first replace those null values using the code below:
        - print(categorical['GENDER'].value_counts())
        - categorical['GENDER'] = categorical['GENDER'].fillna('F')

### Importing libraries and dependencies

In [1]:
import pandas as pd
import numpy as np

### Importing the data

In [2]:
df = pd.read_csv("files_for_activities\learningSet.csv")
df

C:\Users\victo\AppData\Local\Temp\ipykernel_4708\3065485283.py:1: DtypeWarning: Columns (8) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("files_for_activities\learningSet.csv")


ODATEDW OSOURCE  TCODE STATE    ZIP MAILCODE PVASTATE   DOB NOEXCH  \
0         8901     GRI      0    IL  61081                    3712      0   
1         9401     BOA      1    CA  91326                    5202      0   
2         9001     AMH      1    NC  27017                       0      0   
3         8701     BRY      0    CA  95953                    2801      0   
4         8601              0    FL  33176                    2001      0   
...        ...     ...    ...   ...    ...      ...      ...   ...    ...   
95407     9601     ASE      1    AK  99504                       0      0   
95408     9601     DCD      1    TX  77379                    5001      0   
95409     9501     MBC      1    MI  48910                    3801      0   
95410     8601     PRV      0    CA  91320                    4005      0   
95411     8801     MCC      2    NC  28409                    1801      0   

      RECINHSE  ... TARGET_D HPHONE_D RFA_2R RFA_2F RFA_2A MDMAUD_R  MDMAUD_F  \
0               ...      0.0        0      L      4      E        X         X   
1               ...      0.0        0      L      2      G        X         X   
2               ...      0.0        1      L      4      E        X         X   
3               ...      0.0        1      L      4      E        X         X   
4            X  ...      0.0        1      L      2      F        X         X   
...        ...  ...      ...      ...    ...    ...    ...      ...       ...   
95407           ...      0.0        0      L      1      G        X         X   
95408           ...      0.0        1      L      1      F        X         X   
95409           ...      0.0        1      L      3      E        X         X   
95410        X  ...     18.0        1      L      4      F        X         X   
95411        X  ...      0.0        1      L      1      G        C         1   

      MDMAUD_A CLUSTER2 GEOCODE2  
0            X     39.0        C  
1            X      1.0        A  
2            X     60.0        C  
3            X     41.0        C  
4            X     26.0        A  
...        ...      ...      ...  
95407        X     12.0        C  
95408        X      2.0        A  
95409        X     34.0        B  
95410        X     11.0        A  
95411        C     12.0        C  

[95412 rows x 481 columns]

### Splitting categorical features from the dataset

In [3]:
# As we are going to work only with the categorical features, we split off them from the rest of the dataset

categorical = df.select_dtypes(include="object")
categorical

OSOURCE STATE    ZIP MAILCODE PVASTATE NOEXCH RECINHSE RECP3 RECPGVG  \
0         GRI    IL  61081                        0                          
1         BOA    CA  91326                        0                          
2         AMH    NC  27017                        0                          
3         BRY    CA  95953                        0                          
4                FL  33176                        0        X     X           
...       ...   ...    ...      ...      ...    ...      ...   ...     ...   
95407     ASE    AK  99504                        0                          
95408     DCD    TX  77379                        0                          
95409     MBC    MI  48910                        0              X           
95410     PRV    CA  91320                        0        X                 
95411     MCC    NC  28409                        0        X             X   

      RECSWEEP  ... RFA_21 RFA_22 RFA_23 RFA_24 RFA_2R RFA_2A MDMAUD_R  \
0               ...    S4E    S4E    S4E    S4E      L      E        X   
1               ...    N1E    N1E           F1E      L      G        X   
2               ...           S4D    S4D    S3D      L      E        X   
3               ...    A1D    A1D                    L      E        X   
4               ...    A3D    I4E    A3D    A3D      L      F        X   
...        ...  ...    ...    ...    ...    ...    ...    ...      ...   
95407           ...                                  L      G        X   
95408           ...                                  L      F        X   
95409           ...    P1D    P1D                    L      E        X   
95410           ...    A1F    A1F    S2F    S3F      L      F        X   
95411           ...    S3G    S3G    S3G    S3G      L      G        C   

      MDMAUD_F MDMAUD_A GEOCODE2  
0            X        X        C  
1            X        X        A  
2            X        X        C  
3            X        X        C  
4            X        X        A  
...        ...      ...      ...  
95407        X        X        C  
95408        X        X        A  
95409        X        X        B  
95410        X        X        A  
95411        1        C        C  

[95412 rows x 74 columns]

## 1. Check for null values in all the columns

In [4]:
df.isna().sum()/len(df)

ODATEDW     0.000000
OSOURCE     0.000000
TCODE       0.000000
STATE       0.000000
ZIP         0.000000
              ...   
MDMAUD_R    0.000000
MDMAUD_F    0.000000
MDMAUD_A    0.000000
CLUSTER2    0.001383
GEOCODE2    0.001383
Length: 481, dtype: float64

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 95412 entries, 0 to 95411
Columns: 481 entries, ODATEDW to GEOCODE2
dtypes: float64(97), int64(310), object(74)
memory usage: 350.1+ MB


**At first glance, we can see that we have a big dataset with 481 different features, 407 are numerical and 74 are categorical, some of them have nan values or, directly, missing.**<br>

**As we are going to work with categorical features, we are going to check them only**

In [6]:
categorical.isna().sum()/len(df)

OSOURCE     0.000000
STATE       0.000000
ZIP         0.000000
MAILCODE    0.000000
PVASTATE    0.000000
              ...   
RFA_2A      0.000000
MDMAUD_R    0.000000
MDMAUD_F    0.000000
MDMAUD_A    0.000000
GEOCODE2    0.001383
Length: 74, dtype: float64

In [7]:
categorical.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 95412 entries, 0 to 95411
Data columns (total 74 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   OSOURCE   95412 non-null  object
 1   STATE     95412 non-null  object
 2   ZIP       95412 non-null  object
 3   MAILCODE  95412 non-null  object
 4   PVASTATE  95412 non-null  object
 5   NOEXCH    95412 non-null  object
 6   RECINHSE  95412 non-null  object
 7   RECP3     95412 non-null  object
 8   RECPGVG   95412 non-null  object
 9   RECSWEEP  95412 non-null  object
 10  MDMAUD    95412 non-null  object
 11  DOMAIN    95412 non-null  object
 12  CLUSTER   95412 non-null  object
 13  AGEFLAG   95412 non-null  object
 14  HOMEOWNR  95412 non-null  object
 15  CHILD03   95412 non-null  object
 16  CHILD07   95412 non-null  object
 17  CHILD12   95412 non-null  object
 18  CHILD18   95412 non-null  object
 19  GENDER    95412 non-null  object
 20  DATASRCE  95412 non-null  object
 21  SOLP3     95

**We have nans only in one categorical feature, "GEOCODE2"**

## 2. Excluding some features

In [8]:
# Creating a vartiable to store all the features that are going to be deleted

drop_list = []

In [9]:
drop_list.append("OSOURCE")
drop_list.append("ZIP")

In [10]:
drop_list # Sanity check

['OSOURCE', 'ZIP']

## 3. Identify columns that have over 85% missing values

>I am going to create a loop to iterate through all the columns of the categorical data and append the columns that have 85% or more of empty values to the drop_list we created before.

In [11]:
total_columns_dropped = 0  # I put a count to store the number of columns that have more than 85% of empty spaces
for column in categorical:
    categorical[column] = categorical[column].apply(str) # Changing any possible integer or float value to string
    space_count = 0  # This count stores the total nomber of empty spaces per feature
    for value in categorical[column]:
        x = value.isspace()  # Checking if all the string is a space (e.g "   e   " would be False)
        if x == True:
            space_count += 1
    if space_count/len(categorical[column]) >= 0.85:  # Checking the proportion of spaces there are in the feature
        drop_list.append(column)
        total_columns_dropped += 1
    space_count = 0  # reseting the count
print("")
print('\033[1m'"There are {} columns with more than 85% of empty values"'\033[0m'.format(total_columns_dropped))
print("")


There are 30 columns with more than 85% of empty values



>Building the loop, I observed that some of the columns had integers or floats as values even tho the column type was an object and it was giving me troubles (<span style="color: red">AttributeError:</span>  'int' object has no attribute 'isspace') that is why I decided to include code row to transform all the values of those columns into string to avoid it.

In [12]:
drop_list # Sanity check

['OSOURCE',
 'ZIP',
 'MAILCODE',
 'PVASTATE',
 'RECINHSE',
 'RECP3',
 'RECPGVG',
 'RECSWEEP',
 'CHILD03',
 'CHILD07',
 'CHILD12',
 'CHILD18',
 'SOLP3',
 'SOLIH',
 'MAJOR',
 'COLLECT1',
 'VETERANS',
 'BIBLE',
 'CATLG',
 'HOMEE',
 'CDPLAY',
 'STEREO',
 'PCOWNERS',
 'PHOTO',
 'CRAFTS',
 'FISHER',
 'GARDENIN',
 'BOATS',
 'WALKER',
 'KIDSTUFF',
 'CARDS',
 'PLATES']

## 4. Remove those columns from the dataframe

In [13]:
categorical = categorical.drop(columns=drop_list, axis=0)

In [14]:
categorical

STATE NOEXCH MDMAUD DOMAIN CLUSTER AGEFLAG HOMEOWNR GENDER DATASRCE  \
0        IL      0   XXXX     T2      36                       F            
1        CA      0   XXXX     S1      14       E        H      M        3   
2        NC      0   XXXX     R2      43                U      M        3   
3        CA      0   XXXX     R2      44       E        U      F        3   
4        FL      0   XXXX     S2      16       E        H      F        3   
...     ...    ...    ...    ...     ...     ...      ...    ...      ...   
95407    AK      0   XXXX     C2      27                       M            
95408    TX      0   XXXX     C1      24       E        H      M        3   
95409    MI      0   XXXX     C3      30                       M            
95410    CA      0   XXXX     C1      24       E        H      F        2   
95411    NC      0   C1CM     C1      24       E        U      F        3   

      GEOCODE  ... RFA_21 RFA_22 RFA_23 RFA_24 RFA_2R RFA_2A MDMAUD_R  \
0              ...    S4E    S4E    S4E    S4E      L      E        X   
1           2  ...    N1E    N1E           F1E      L      G        X   
2              ...           S4D    S4D    S3D      L      E        X   
3              ...    A1D    A1D                    L      E        X   
4              ...    A3D    I4E    A3D    A3D      L      F        X   
...       ...  ...    ...    ...    ...    ...    ...    ...      ...   
95407          ...                                  L      G        X   
95408          ...                                  L      F        X   
95409          ...    P1D    P1D                    L      E        X   
95410       4  ...    A1F    A1F    S2F    S3F      L      F        X   
95411       3  ...    S3G    S3G    S3G    S3G      L      G        C   

      MDMAUD_F MDMAUD_A GEOCODE2  
0            X        X        C  
1            X        X        A  
2            X        X        C  
3            X        X        C  
4            X        X        A  
...        ...      ...      ...  
95407        X        X        C  
95408        X        X        A  
95409        X        X        B  
95410        X        X        A  
95411        1        C        C  

[95412 rows x 42 columns]

## 5. Reduce the number of categories in the column GENDER

In [15]:
# We will replace the null values of column 'GENDER' with 'F' as it was said in the Instructions

categorical['GENDER'].value_counts()

F    51277
M    39094
      2957
U     1715
J      365
C        2
A        2
Name: GENDER, dtype: int64

In [16]:
categorical['GENDER'] = categorical['GENDER'].fillna('F')

In [17]:
categorical['GENDER'].value_counts() #Sanity check

F    51277
M    39094
      2957
U     1715
J      365
C        2
A        2
Name: GENDER, dtype: int64

**This does not produce anything because there are no nans but empty spaces**

In [18]:
# Replacing empty spaces for 'F' and the rest for 'other' and keeping 'M'

categorical['GENDER'] = categorical['GENDER'].replace(' ', 'F').replace(['U', 'J', 'C', 'A'], "other")

In [19]:
categorical['GENDER'].value_counts() # Sanity check

F        54234
M        39094
other     2084
Name: GENDER, dtype: int64